feat: ujust label-external-drives

pxlkng opened the pull request:
Closes #891.

~DRAFT: Still need to figure out the underlying tooling issue with the labels not applying.~
Fixed by 4bf9b05.

pxlkng committed:
feat: ujust label-external-drives

pxlkng committed:
style: remove unnecessary spaces

pxlkng committed:
style: use single quotes where they are not required for variable expansion

pxlkng committed:
fix: labeling path issue

The underlying labeling issues seems to have been due to the real path being "/run/media" and "/media" just being a symlink which apparently doesn't propagate to fcontext rules.

pxlkng committed:
fix: broken directory ownership check

pxlkng committed:
fix: prevent globbing and word splitting on ownership check

pxlkng committed:
style: add missing "is"

RoyalOughtness commented on files/justfiles/utilities.just:
```diff
@@ -253,3 +253,51 @@ uninstall-docker:
+    done
+
+    echo '' # blank space
+    echo 'Done.'
```
echo -e '\nDone.'

pxlkng replied:
Fixed by bd8809e.

RoyalOughtness commented on files/justfiles/utilities.just:
```diff
@@ -253,3 +253,51 @@ uninstall-docker:
+        fi
+        echo "Labeling drive mounted on $dir with $selabel"
+        dir=${dir%*/} # remove trailing slash
+        if ! run0 sh -c "semanage fcontext -a -t $selabel \"$dir(/.*)?\" && restorecon -r $dir"; then
```
I think if you use ' for the outer quotes you can avoid the need to escape the inner quotes

pxlkng replied:
It should be the other way around if I'm not mistaken.
If I use single quotes for the outer ones, it will stop the `$dir` variable from expanding and pass it literally to the command string.
But you're still right, I can use single quotes for the _inner_ ones to avoid having to escape.
Will correct that.

I just now come to think, shouldn't the second `$dir` (for restorecon) be put in single quotes as well to avoid any interpretation or word splitting of its contents? Couldn't I just name a folder `XY && malicious-binary` and that would expand to `[...] && restorecon -r XY && malicious-binary` and get executed as root, with how it is currently implemented?

RoyalOughtness replied:
yeah and we could also add a test to ensure a dir with that name breaks it

pxlkng replied:
Wouldn't single quotes around `$dir` force everything in the variable to not be interpreted or expanded, except if it itself contains another single quote?

If we check manually, against which specific characters should we do so? Only single quotes, as everything else should already be contained by them being around `$dir`?

RoyalOughtness replied:
I believe that's right but I don't remember for certain off the top of my head

RoyalOughtness replied:
worth double checking

pxlkng replied:
Yeah, looks like single quotes prevent anything from being interpreted except themselves. (which makes sense)

Will do the single quotes and implement a security check against the `$dir` containing them and skip with a warning if it does.

pxlkng replied:
Implemented in 2206b3c & 1a7f298.

RoyalOughtness commented on files/justfiles/utilities.just:
```diff
@@ -253,3 +253,51 @@ uninstall-docker:
+            continue
+        fi
+        detected_fs="$(df -T "$dir" | awk '{print $2}' | sed -n 2p)"
+        if ! [[ "${supported_fs[*]}" =~ "$detected_fs" ]]; then
```
shouldn't this be a precise comparison as opposed to a loose one?

pxlkng replied:
Corrected by c71e66f.

pxlkng committed:
style: remove unnecessary blank line and use inline-newline instead

pxlkng committed:
fix: do precise instead of loose comparison for fs compat check

pxlkng committed:
style: consistently put dir path in quotes for echoed messages

for weird cases and better readability (could look weird if directories have spaces or similar for some reason, holds the path together as one)

pxlkng committed:
fix: use single quotes around dir variable to avoid expansion

pxlkng committed:
fix: implement security check to avoid command execution

RoyalOughtness commented on files/justfiles/utilities.just:
```diff
@@ -253,3 +253,54 @@ uninstall-docker:
+
+    for dir in "$media_path"*/
+    do
+        echo '' # blank space
```
(nit) `echo` on its own can do this I believe

RoyalOughtness replied:
also we probably don't need the comment explaining it :smile:

pxlkng replied:
I just inherited that from another script lol, but true.

And I'll just believe you and do `echo` without quotes as I can't test that on my phone right now.

pxlkng replied:
789604b 👍

pxlkng committed:
style: optimize echo and remove unnecessary comment

ShadowSlayer1441 commented on files/justfiles/utilities.just:
```diff
@@ -253,3 +253,54 @@ uninstall-docker:
+    echo "Warning: This will label your drives mounted on \"$media_path\" such that they are treated as an extension to your home directory. (You can select which ones to label)"
+    echo 'If you do not want Trivalent to have access to these drives, do not proceed.'
+    read -p 'Are you sure you want to do this? [y/N] ' warning_proceed
+    if ! [[ "$warning_proceed" == [Yy]* ]]; then
```
This should be `if [[ "$warning_proceed" != [Nn]* ]]; then` 
This makes the default (if you press enter) exiting and is more clear in general.

pxlkng replied:
The default already is exiting. (notice the negating `!`)

Your snippet would reverse the logic, since you not only changed the place of the negating exclamation mark (which is a fine suggestion) but also changed the y-check to an n-check.

This would make inputting `y` exit and entering `n` continue.

ShadowSlayer1441 commented on files/justfiles/utilities.just:
```diff
@@ -253,3 +253,54 @@ uninstall-docker:
+    do
+        echo
+        if [[ "$dir" =~ "'" ]]; then
+            echo "WARNING: Single quotes are forbidden to avoid command execution, skipping drive \"$dir\"..."
```
echo "WARNING: Single quotes in drive labels are forbidden to avoid command execution, skipping drive \"$dir\"..."

pxlkng replied:
Improved by 7216aca.

pxlkng committed:
style: clarify warning on single quotes

RoyalOughtness committed:
Merge branch 'live' into feat/ujust-label-external-drives
